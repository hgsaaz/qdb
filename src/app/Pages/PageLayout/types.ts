export interface UserState {
  username: string;
  user: UserInfo | null;
  error?: UserErrorType | null;
  loading: Boolean;
}

export interface AddressInfo {
  city: string;
  geo: { lat: number; lng: number };
  street: string;
  suite: string;
  zipcode: string;
}

export interface CompanyInfo {
  bs: string;
  catchPhrase: string;
  name: string;
}

export interface UserInfo {
  email: string;
  id: string;
  name: string;
  phone: string;
  website: string;
  address: AddressInfo;
  company: CompanyInfo;
}

export enum UserErrorType {
  RESPONSE_ERROR = 1,
  USER_NOT_FOUND = 2,
}

export type ContainerState = UserState;
