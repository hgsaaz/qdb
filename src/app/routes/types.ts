export interface RoutesList {
  path: string;
  component: string;
  type: string;
}
