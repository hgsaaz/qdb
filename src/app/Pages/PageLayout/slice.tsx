import { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from 'utils/@reduxjs/toolkit';
import { ContainerState, UserErrorType, UserInfo } from './types';

export const initialState: ContainerState = {
  username: '',
  user: null,
  error: null,
  loading: true,
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    saveUser(state, action: PayloadAction<UserInfo>) {
      state.user = action.payload;
      state.loading = false;
    },
    loadUser(state, action: PayloadAction<string>) {
      state.username = action.payload;
      state.error = null;
    },
    fetchUserError(state, action: PayloadAction<UserErrorType>) {
      state.error = action.payload;
      state.loading = false;
    },
  },
});

export const { actions, reducer, name: sliceKey } = userSlice;
