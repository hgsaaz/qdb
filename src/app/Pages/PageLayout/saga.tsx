import { call, put, select, takeLatest } from 'redux-saga/effects';
import { request } from 'utils/request';
import { selectUsername, selectUser } from './selectors';
import { actions } from './slice';
import { UserErrorType, UserInfo } from './types';
import { message } from 'antd';

/**
 * Github repos request/response handler
 */
export function* getUser() {
  // Select username from store
  const username: string = yield select(selectUsername);
  const user: UserInfo = yield select(selectUser);
  if (username.length === 0) {
    return;
  }
  const requestURL = `https://jsonplaceholder.typicode.com/users/${username}`;

  try {
    // Call our request helper (see 'utils/request')
    if (user === null) {
      const userdetails = yield call(request, requestURL);
      yield put(actions.saveUser(userdetails));
    }
  } catch (err) {
    if (err.response?.status === 404) {
      yield put(actions.repoError(UserErrorType.USER_NOT_FOUND));
      message.error('This is an error message');
    }
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* userSaga() {
  // Watches for loadRepos actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(actions.loadUser.type, getUser);
}
