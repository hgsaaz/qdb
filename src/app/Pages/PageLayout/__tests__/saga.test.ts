import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '../slice';

import { userSaga, getUser } from '../saga';
import { UserErrorType } from '../types';

describe('getUser Saga', () => {
  let username: any;
  let user: any;
  let getUserIterator: ReturnType<typeof getUser>;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach

  beforeEach(() => {
    getUserIterator = getUser();
    // const selectUsernameDescriptor = getUserIterator.next().value;
    // expect(selectUsernameDescriptor).toMatchSnapshot();
    // const selectUserDescriptor = getUserIterator.next().value;
    // expect(selectUserDescriptor).toMatchSnapshot();
  });

  it('should return error if username is empty', () => {
    // username = 'saasa';
    // const usernameErrorDescriptor = getUserIterator.next(username).value;
    // expect(usernameErrorDescriptor).toEqual(
    //   put(slice.actions.fetchUserError(UserErrorType.USERNAME_EMPTY)),
    // );
    // const iteration = getUserIterator.next();
    // expect(iteration.done).toBe(true);
  });

  it('should dispatch the saveUser action if it requests the data successfully', () => {
    username = '1';
    user = null;

    // console.log(getUserIterator.next().value);
    // console.log(getUserIterator.next().value);
    // const clone = getUserIterator.clone();
    // console.log(getUserIterator.next().value);
    // console.log(getUserIterator.next(user).value);
    // const putDescriptor = getUserIterator.next(user).value;
    // expect(putDescriptor).toEqual(put(slice.actions.saveUser(user)));
  });

  it('should dispatch the user not found error', () => {
    // username = 'test';
    // const putDescriptor = getUserIterator.throw({ response: { status: 404 } })
    //   .value;
    // expect(putDescriptor).toEqual(
    //   put(slice.actions.fetchUserError(UserErrorType.USER_NOT_FOUND)),
    //);
  });
});

describe('userSaga Saga', () => {
  const userSaga1 = userSaga();
  it('should start task to watch for loadUser action', () => {
    const takeLatestDescriptor = userSaga1.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.actions.loadUser.type, getUser),
    );
  });
});
