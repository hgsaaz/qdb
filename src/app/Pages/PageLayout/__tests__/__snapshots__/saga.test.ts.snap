// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUser Saga should return error if username is empty 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getUser Saga should return error if username is empty 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
